// neon-init/dist/lib/auth.js
import { log } from "@clack/prompts";
import { existsSync, readFileSync } from "node:fs";
import { resolve } from "node:path";
import { execa } from "execa";
//#region src/lib/auth.ts
/**
* Ensures neonctl is authenticated by running a command that triggers auth if needed
* This will automatically start the OAuth flow if the user isn't already authenticated
*/
async function ensureNeonctlAuth() {
	if (await getNeonctlAccessToken()) return true;
	try {
		await execa("npx", [
			"-y",
			"neonctl",
			"me",
			"--no-analytics"
		], {
			stdio: "inherit",
			env: {
				...process.env,
				CI: void 0
			}
		});
		return true;
	} catch (error) {
		const msg = error instanceof Error ? error.message : "Unknown error";
		if (msg.includes("interactive auth") || msg.includes("CI")) log.error("Auth requires an interactive terminal. Run neon-init in your system terminal (outside the chat) to sign in.");
		else log.error(`Authentication failed: ${msg}`);
		return false;
	}
}
/**
* Gets the OAuth access token from neonctl's stored credentials
*/
async function getNeonctlAccessToken() {
	try {
		const homeDir = process.env.HOME || process.env.USERPROFILE;
		if (!homeDir) return null;
		const credentialsPath = resolve(homeDir, ".config", "neonctl", "credentials.json");
		if (!existsSync(credentialsPath)) return null;
		return JSON.parse(readFileSync(credentialsPath, "utf-8")).access_token || null;
	} catch {
		return null;
	}
}
/**
* Creates an API key using the Neon API with the OAuth token from neonctl
*/
async function createApiKeyFromNeonctl() {
	try {
		const accessToken = await getNeonctlAccessToken();
		if (!accessToken) {
			log.error("Could not find OAuth token from neonctl");
			return null;
		}
		const keyName = `neonctl-init-${(/* @__PURE__ */ new Date()).toISOString().replace(/[:.]/g, "-").slice(0, -5)}`;
		const response = await fetch("https://console.neon.tech/api/v2/api_keys", {
			method: "POST",
			headers: {
				Authorization: `Bearer ${accessToken}`,
				"Content-Type": "application/json"
			},
			body: JSON.stringify({ key_name: keyName })
		});
		if (!response.ok) {
			const errorText = await response.text();
			log.error(`Failed to create API key: ${response.status} ${errorText}`);
			return null;
		}
		return (await response.json()).key || null;
	} catch (error) {
		log.error(`Failed to create API key: ${error instanceof Error ? error.message : "Unknown error"}`);
		return null;
	}
}
//#endregion
export { createApiKeyFromNeonctl, ensureNeonctlAuth };
